project_task_stage_change_restriction/models/project_task_stage.py

```diff
@@ -0,0 +1,36 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProjectTaskStage(models.Model):
+    _inherit = "project.task.type"
+
+    assigned_only = fields.Boolean()
+    project_manager_only = fields.Boolean(string="Project Manager")
+    allowed_group_ids = fields.Many2many(
+        "res.groups",
+        "project_task_stage_group_rel",
+        "stage_id",
+        "group_id",
+        string="Group Members",
+    )
+
+    def has_restrictions(self):
+        """Return True if the stage has at least one restriction."""
+        self.ensure_one()
+        return bool(
+            self.assigned_only or self.project_manager_only or self.allowed_group_ids
+        )
+
+    def _user_passes_group_rule(self, user):
+        """
+        Return True if the stage has group restrictions
+        and the user belongs to at least one of the allowed groups.
+        If no groups are specified, this rule is not applied.
+        """
+        self.ensure_one()
+        if not self.allowed_group_ids:
+            return False
+        return bool(self.allowed_group_ids & user.groups_id)
```

Comment:
You have all the check logic in the `project.task` model. So it's reasonable to have all the functions there as well.
NB: if there are more that one group in this field, user can be a member of ANY of them. Not ALL.